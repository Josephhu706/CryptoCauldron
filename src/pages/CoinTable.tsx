import React, { useEffect, useState } from "react";
import "../App.css";
import { fetchCrypto, Coin } from "../config/API";
import { ThemeProvider, createTheme } from "@mui/material/styles";
import { Container } from "@mui/system";
import { CssBaseline } from "@mui/material/";
import Table from "@mui/material/Table";
import TableBody from "@mui/material/TableBody";
import TableCell from "@mui/material/TableCell";
import TableContainer from "@mui/material/TableContainer";
import TableHead from "@mui/material/TableHead";
import CircularProgress from "@mui/material/CircularProgress";
import TableRow from "@mui/material/TableRow";
import Pagination from "@mui/material/Pagination";
import callError from "../assets/images/call-error.svg";
import emptySearch from "../assets/images/empty-search.svg";
import SearchInput from "../components/SearchInput";
import MotionWrap from "../wrapper/MotionWrap";
import "../styles/CoinTableStyles.scss";
import { useNavigate } from "react-router-dom";

const darkTheme = createTheme({
  palette: {
    mode: "dark",
  },
});

const CoinTable = () => {
  const [coins, setCoins] = useState<Coin[]>([]);
  const [search, setSearch] = useState<string>("");
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string>("");
  const [page, setPage] = useState<number>(1);
  const tableHead = ["Coin", "Price", "Volume", "24hr Change", "Market Cap"];

  const navigate = useNavigate();

  useEffect(() => {
    getCoins();
  }, []);

  const getCoins = async () => {
    setLoading(true);
    try {
      const newCoins = await fetchCrypto();
      setCoins(newCoins);
    } catch (error: any) {
      setError(error);
    }
    setLoading(false);
  };

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearch(event.target.value);
  };

  const filterCoins = () => {
    return coins.filter(
      (coin) =>
        coin.name.toLowerCase().includes(search.toLowerCase()) ||
        coin.symbol.toLowerCase().includes(search.toLowerCase())
    );
  };

  return (
    <div data-testid="coinTable">
      <ThemeProvider theme={darkTheme}>
        <CssBaseline />
        <Container className="tableContainer">
          <SearchInput handleChange={handleChange} />
          <TableContainer>
            {loading ? (
              <CircularProgress data-testid="loading" />
            ) : error ? (
              <div data-testid="fetchError">
                <img
                  src={callError}
                  style={{ height: "20em" }}
                  alt="fetchError"
                />
              </div>
            ) : filterCoins().length === 0 ? (
              <div data-testid="emptySearch">
                <img
                  src={emptySearch}
                  style={{ height: "20em" }}
                  alt="emptySearch"
                />
              </div>
            ) : (
              <Table data-testid="fetchSuccessful">
                <TableHead className="tableHead">
                  <TableRow>
                    {tableHead.map((head, index) => (
                      <TableCell
                        align={head === "Coin" ? "left" : "right"}
                        key={index}
                        className="tableHeadCell"
                      >
                        {head}
                      </TableCell>
                    ))}
                  </TableRow>
                </TableHead>
                <TableBody id="tableBody">
                  {filterCoins()
                    .slice((page - 1) * 10, (page - 1) * 10 + 10)
                    .map((row, index) => {
                      return (
                        <TableRow
                          className="tableRow"
                          key={index}
                          onClick={() => navigate(`/coins/${row.id}`)}
                        >
                          <TableCell scope="row" className="tableCellCoin">
                            <img
                              src={row.image}
                              className="tableCoinIcon"
                              alt={row.name}
                            />
                            <div className="tableCoinContent">
                              <span
                                style={{
                                  textTransform: "uppercase",
                                  fontSize: 22,
                                }}
                              >
                                {row.symbol}
                              </span>
                              <span style={{ color: "darkgrey" }}>
                                {row.name}
                              </span>
                            </div>
                          </TableCell>

                          <TableCell align="right">
                            ${row.current_price.toFixed(2)} AUD
                          </TableCell>

                          <TableCell align="right">
                            ${row.total_volume.toLocaleString()}
                          </TableCell>

                          <TableCell align="right">
                            {row.price_change_24h < 0 ? (
                              <p style={{ color: "red" }}>
                                {row.price_change_24h.toFixed(2)}%
                              </p>
                            ) : (
                              <p style={{ color: "limegreen" }}>
                                {row.price_change_24h.toFixed(2)}%
                              </p>
                            )}
                          </TableCell>

                          <TableCell align="right">
                            ${row.market_cap.toLocaleString()}
                          </TableCell>
                        </TableRow>
                      );
                    })}
                </TableBody>
              </Table>
            )}
            {!error && (
              <Pagination
                data-testid="pagination"
                style={{
                  padding: 20,
                  width: "100%",
                  display: "flex",
                  justifyContent: "center",
                }}
                count={parseInt((filterCoins().length / 10).toFixed(0))}
                color="primary"
                onChange={(event, value) => {
                  setPage(value);
                  window.scroll(0, 450);
                }}
              />
            )}
          </TableContainer>
        </Container>
      </ThemeProvider>
    </div>
  );
};

export default MotionWrap(CoinTable);
