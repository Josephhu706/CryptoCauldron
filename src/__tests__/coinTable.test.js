import { render, screen, cleanup, waitForElement, getByTestId, fireEvent, } from '@testing-library/react'
import CoinTable from '../pages/CoinTable'
import '@testing-library/jest-dom'
import axiosMock from 'axios'
import renderer from 'react-test-renderer'


// //makes sure each test is starting from the same starting point
afterEach(()=>{
    cleanup()
})

jest.mock('axios');

//test if coin table component is rendered
test("table should be rendered", () => {
    render(<CoinTable />);
    const coinTable = screen.getByTestId("coinTable");
    expect(coinTable).toBeInTheDocument()
  });

  //test if logo is rendered
test("table logo should be rendered", () => {
    render(<CoinTable />);
    const tableLogo = screen.getByTestId("tableLogo");
    expect(tableLogo).toBeInTheDocument()
});

//test if search bar is rendered
test("search input should be rendered", () => {
    render(<CoinTable />);
    const inputEl= screen.getByTestId("search");
    expect(inputEl).toBeInTheDocument();
});

test("search error should be not be visible", () => {
  render(<CoinTable />);
  const errorEl = screen.queryByTestId('emptySearch');
  expect(errorEl).not.toBeInTheDocument()
});

//test fetching API
test("Coins should be rendered after fetching", async () => {
  axiosMock.get.mockResolvedValueOnce({data: [            
    {"id": "bitcoin",
    "symbol": "btc",
    "name": "Bitcoin",
    "image": "https://assets.coingecko.com/coins/images/1/large/bitcoin.png?1547033579",
    "current_price": 39161,
    "market_cap": 746975620874,
    "market_cap_rank": 1,
    "fully_diluted_valuation": 822780677096,
    "total_volume": 37007140684,
    "high_24h": 41702,
    "low_24h": 38967,
    "price_change_24h": -2356.366268313068,
    "price_change_percentage_24h": -5.67563,
    "market_cap_change_24h": -44140206844.06592,
    "market_cap_change_percentage_24h": -5.57949,
    "circulating_supply": 19065212,
    "total_supply": 21000000,
    "max_supply": 21000000,
    "ath": 93482,
    "ath_change_percentage": -57.95187,
    "ath_date": "2021-11-10T14:24:11.849Z",
    "atl": 72.61,
    "atl_change_percentage": 54035.65854,
    "atl_date": "2013-07-05T00:00:00.000Z",
    "roi": null,
    "last_updated": "2022-06-12T03:58:19.387Z"
    }
]})
  render(<CoinTable />);
  const loading= screen.getByTestId("loading")
  expect(loading).toBeInTheDocument();
  const tableContent = await screen.findByText('Bitcoin')
  expect(tableContent).toBeInTheDocument();
});

//test for not fetching data error screen
test("Render Error Message if Fetch Fails", async () => {
  axiosMock.get.mockRejectedValue({data: {'error': 'error'}})
  render(<CoinTable />);
  const errorMsg = await screen.findByTestId('fetchError')
  expect(errorMsg).toBeInTheDocument();
});


// //test for route pagination?

// //test for filtering?
// test('filtering should work', ()=>{
//   // const coins = {content:[ {"id": "bitcoin","symbol": "btc", "name": "Bitcoin"},{"id": 'ethereum', "symbol": 'eth', "name": 'Ethereum'}]}
//   const wrapper = shallow(<CoinTable/>);
//   const instance = wrapper.instance()
//   const func = instance.filterCoins()
//   console.log(instance.debug())
//   const mockedChangeEvent = { target: { value: 'Bitcoin' } };
//   expect(wrapper.state('search')).toBe('');
//   wrapper.find('#searchCoin').simulate('change', mockedChangeEvent);
//   expect(wrapper.state('search')).toBe('Bitcoin');
//   expect(wrapper.find('#tablebody').children()).toHaveLength(1);
// })

test("search input should change", () => {
  render(<CoinTable />);
  const searchInputEl = screen.getByTestId('search').querySelector('input');
  //we have a dummy test value
  const testValue = 'test'
  //if we fire an onchange vent for the input 
  fireEvent.change(searchInputEl, {target: {value: testValue}})
  expect(searchInputEl.value).toBe(testValue);
});

test("Pagination should be rendered", () => {
  render(<CoinTable />);
  const pagination= screen.getByTestId("pagination");
  expect(pagination).toBeInTheDocument();
});

test('matches snapshot', () =>{
  const tree = renderer.create(<CoinTable/>).toJSON()
  expect(tree).toMatchSnapshot();
})