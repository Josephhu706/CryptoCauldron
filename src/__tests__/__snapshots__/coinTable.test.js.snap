// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  style={Object {}}
>
  <div
    data-testid="coinTable"
  >
    <div
      className="MuiContainer-root MuiContainer-maxWidthLg tableContainer css-1oqqzyl-MuiContainer-root"
    >
      <div>
        <div
          data-testid="tableLogo"
        >
          <img
            alt="logo"
            className="searchLogo"
            src="logo.svg"
          />
          <h5
            className="MuiTypography-root MuiTypography-h5 css-1m67wze-MuiTypography-root"
            style={
              Object {
                "fontFamily": "Montserrat",
                "margin": 18,
              }
            }
          >
            Cook up Coins in the Crypto Cauldron
          </h5>
        </div>
        <div
          className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          data-testid="search"
          style={
            Object {
              "marginBottom": 20,
              "width": "80%",
            }
          }
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-rb5gc9-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor="outlined-basic"
            id="outlined-basic-label"
          >
            Search for Coin Name or Symbol
          </label>
          <div
            className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-px39cz-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiOutlinedInput-input MuiInputBase-input css-p51h6s-MuiInputBase-input-MuiOutlinedInput-input"
              data-testid="content-input"
              disabled={false}
              id="outlined-basic"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-9425fu-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-1ftyaf0"
              >
                <span>
                  Search for Coin Name or Symbol
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        className="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
      >
        <div
          data-testid="emptySearch"
        >
          <img
            alt="emptySearch"
            src="empty-search.svg"
            style={
              Object {
                "height": "20em",
              }
            }
          />
        </div>
        <nav
          aria-label="pagination navigation"
          className="MuiPagination-root MuiPagination-text css-1oj2twp-MuiPagination-root"
          data-testid="pagination"
          style={
            Object {
              "display": "flex",
              "justifyContent": "center",
              "padding": 20,
              "width": "100%",
            }
          }
        >
          <ul
            className="MuiPagination-ul css-wjh20t-MuiPagination-ul"
          >
            <li>
              <button
                aria-label="Go to previous page"
                className="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-textPrimary Mui-disabled MuiPaginationItem-previousNext css-rij76b-MuiButtonBase-root-MuiPaginationItem-root"
                disabled={true}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                  data-testid="NavigateBeforeIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                  />
                </svg>
              </button>
            </li>
            <li>
              <button
                aria-label="Go to next page"
                className="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular MuiPaginationItem-textPrimary Mui-disabled MuiPaginationItem-previousNext css-rij76b-MuiButtonBase-root-MuiPaginationItem-root"
                disabled={true}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={-1}
                type="button"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
                  data-testid="NavigateNextIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                  />
                </svg>
              </button>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </div>
</div>
`;
